libvips/create/sdf.c

```diff
@@ -0,0 +1,396 @@
+			return -1;
+		}
+
+		sdf->a = (double *) sdf->a_area->data;
+		sdf->b = (double *) sdf->b_area->data;
+		sdf->point = vips_sdf_box;
+	}
+	else if (g_str_equal(sdf->name, "rounded-box")) {
+		if (!vips_object_argument_isset(object, "a") ||
+			!vips_object_argument_isset(object, "b")) {
+			vips_error(class->nickname, "%s",
+				_("rounded-box needs a, b to be set"));
+			return -1;
+		}
+		if (sdf->a_area->n != 2 ||
+			sdf->b_area->n != 2) {
+			vips_error(class->nickname, "%s",
+				_("rounded-box needs 2 values for a, b"));
+			return -1;
+		}
+		if (sdf->corners_area->n != 4) {
+			vips_error(class->nickname, "%s",
+				_("rounded-box needs 4 values for corners"));
+			return -1;
+		}
+
+		sdf->a = (double *) sdf->a_area->data;
+		sdf->b = (double *) sdf->b_area->data;
+		sdf->corners = (double *) sdf->corners_area->data;
+		sdf->point = vips_sdf_rounded_box;
+	}
+	else if (g_str_equal(sdf->name, "line")) {
+		if (!vips_object_argument_isset(object, "a") ||
+			!vips_object_argument_isset(object, "b")) {
+			vips_error(class->nickname, "%s",
+				_("line needs sx, sy to be set"));
+			return -1;
+		}
+		if (sdf->a_area->n != 2 ||
+			sdf->b_area->n != 2) {
+			vips_error(class->nickname, "%s",
+				_("line needs 2 values for a, b"));
+			return -1;
+		}
+
+		sdf->a = (double *) sdf->a_area->data;
+		sdf->b = (double *) sdf->b_area->data;
+		sdf->point = vips_sdf_line;
+	}
+	else {
+		vips_error(class->nickname, _("unknown SDF %s"), sdf->name);
+		return -1;
+	}
+
+	if (sdf->a &&
+		sdf->b) {
+		// centre
+		sdf->cx = (sdf->a[0] + sdf->b[0]) / 2.0;
+		sdf->cy = (sdf->a[1] + sdf->b[1]) / 2.0;
+
+		// diffetrence
+		sdf->dx = sdf->b[0] - sdf->a[0];
+		sdf->dy = sdf->b[1] - sdf->a[1];
+
+		// half size
+		sdf->sx = sdf->dx / 2.0;
+		sdf->sy = sdf->dy / 2.0;
+	}
+
+	vips_image_init_fields(create->out,
+		sdf->width, sdf->height, 1,
+		VIPS_FORMAT_FLOAT, VIPS_CODING_NONE, VIPS_INTERPRETATION_B_W, 1.0, 1.0);
+	if (vips_image_pipelinev(create->out, VIPS_DEMAND_STYLE_ANY, NULL) ||
+		vips_image_generate(create->out, NULL, vips_sdf_gen, NULL, sdf, NULL))
+		return -1;
+
+	return 0;
+}
+
+static void
+vips_sdf_class_init(VipsSdfClass *class)
+{
+	GObjectClass *gobject_class = G_OBJECT_CLASS(class);
+	VipsObjectClass *vobject_class = VIPS_OBJECT_CLASS(class);
+
+	VIPS_DEBUG_MSG("vips_sdf_class_init\n");
+
+	gobject_class->set_property = vips_object_set_property;
+	gobject_class->get_property = vips_object_get_property;
+
+	vobject_class->nickname = "sdf";
+	vobject_class->description = _("create an SDF image");
+	vobject_class->build = vips_sdf_build;
+
+	VIPS_ARG_INT(class, "width", 2,
+		_("Width"),
+		_("Image width in pixels"),
+		VIPS_ARGUMENT_REQUIRED_INPUT,
+		G_STRUCT_OFFSET(VipsSdf, width),
+		1, VIPS_MAX_COORD, 1);
+
+	VIPS_ARG_INT(class, "height", 3,
+		_("Height"),
+		_("Image height in pixels"),
+		VIPS_ARGUMENT_REQUIRED_INPUT,
+		G_STRUCT_OFFSET(VipsSdf, height),
+		1, VIPS_MAX_COORD, 1);
+
+	VIPS_ARG_STRING(class, "name", 6,
```

Comment:
A `VIPS_ARG_ENUM` would probably be neater for strongly typed language bindings. For example, in NetVips (using C# 12 and later), this would result in:
```csharp
using var image = Image.Sdf(1000, 1000, Enums.Sdf.RoundedBox, a: [200, 100], b: [500, 200], corners: [0, 0, 0, 50]);
```
instead of:
```csharp
using var image = Image.Sdf(1000, 1000, "rounded-box", a: [200, 100], b: [500, 200], corners: [0, 0, 0, 50]);
```

Reply:
Yes, I was unsure about that. We could potentially end up with a lot of these, so I thought maybe a string was easier.

Reply:
I changed it an enum ... thanks!